File: C++/RESTORE.cpp
// 전반적인 풀이 방법은 생각을 잘 했으나, 바로 전 단어만 알고도 caching이 가능하도록
// 하려면 다른 단어에 포함되는 단어들을 다 제거하고 시작해야 한다.
// 이 아이디어 없이 하려다 보니 함수가 많이 복잡해졌었다.
// 언제쯤이면 이런 아이디어까지 생각해 내 빈틈없이 문제를 풀 수 있을까....

#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <cstring>

using namespace std;

string combine(int, int);
string reconstruct(int, int, int);

vector<string> words;
int memo[(1<<16) + 1][15];

int main(){
    int i,j,k,l;
    vector<string> tmp;
    for(i=0, scanf("%d", &j); i<j; i++){
        words.clear();
        tmp.clear();
        memset(memo, -1, sizeof(memo));
        for(l=0,scanf("%d", &k); l<k; l++){
            string word;
            cin >> word;
            words.push_back(word);
        }

        int to_be_deleted[15];
        memset(to_be_deleted, -1, sizeof(to_be_deleted));
        int idx = 0;
        for(k=0; k<words.size(); k++){
            for(l=k+1; l<words.size(); l++){
                if(words[k].find(words[l]) != string::npos){
                    to_be_deleted[l] = 1;
                }
                else if(words[l].find(words[k]) != string::npos){
                    to_be_deleted[k] = 1;
                }
            }
        }

        for(k=0; k<words.size(); k++){
            if(to_be_deleted[k] == -1){
                tmp.push_back(words[k]);
            }
        }
        words = tmp;
        std::cout << reconstruct(1<<15, -1, words.size()-1) << std::endl;
    }
    return 0;
}

string combine(int idx1, int idx2){
    if(idx1 == -1)
        return words[idx2];
    int interleaving = 0;
    string tmp = words[idx1];
    // reverse(tmp.begin(), tmp.end());
    int len = words[idx1].length();
    for(int i=0; i< min(words[idx1].size(), words[idx2].size()); i++){
        if(words[idx1].substr(len - i - 1) == words[idx2].substr(0, i+1)){
            interleaving = i + 1;
        }
    }
    return words[idx2].substr(interleaving);
}

int get_min_len(int chosen, int prev_word_idx, int level){
    
    int& result = prev_word_idx != -1? memo[chosen][prev_word_idx]: memo[1 << 16][0];
    if(result != -1)
        return result;
    if(level == -1)
        return 0;
    result = 600;
    for(int i=0; i<words.size(); i++){
        if((chosen & (1 << i)) != (1 << i)){
            int modified_chosen = chosen | (1 << i);
            string combined = combine(prev_word_idx, i);
            result = min(result, (int)combined.size() + get_min_len(modified_chosen, i, level -1));
        }
    }
    return result;
}

string reconstruct(int chosen, int prev_word_idx, int level){
    if(level == -1)
        return "";
    int target = get_min_len(chosen, prev_word_idx, level);
    for(int i=0; i< words.size(); i++){
        if((chosen & (1 << i)) != (1 << i)){
            int modified_chosen = chosen | (1 << i);
            string combined = combine(prev_word_idx, i);
            if(target == combined.size() + get_min_len(modified_chosen, i, level - 1)){
                return combined + reconstruct(modified_chosen, i, level-1);
            }
        }
    }
}

File: Python/RESTORE.py
## 알고리즘은 맞는데, 제출하면 RTE 에러가 뜬다.
## 나중에 시간 나면 해결해보자, 일단은 c++로 포팅하여 제출했다.

import sys

global memo
global memo2
global no_of_word

def combine(w1, w2):
    if w1 == -1:
        return words[w2]
    if memo2[w1][w2] != '':
        return memo2[w1][w2]
    interleaving = 0
    for i in range(min(len(words[w1]), len(words[w2]))):
        if words[w1][-(i+1):] == words[w2][0:i+1]:
            interleaving = i + 1
        else:
            break
    memo2[w1][w2] = words[w2][interleaving:]
    return memo2[w1][w2]

def get_min_comb(chosen, prev_word_idx, level):
    prev_word = ""
    if prev_word_idx != -1:
        if memo[chosen][prev_word_idx] != '':
            return memo[chosen][prev_word_idx];
    if level == -1:
        return 0
    if prev_word_idx == -1:
        prev_word = ""
    else:
        prev_word = words[prev_word_idx]
    result = 600;
    for i in range(no_of_word):
        if chosen & 1 << i != 1 << i:
            modified_chosen = chosen | 1 << i;
            combined = combine(prev_word_idx, i)
            result = min(result, len(combined) + get_min_comb(modified_chosen, i, level -1))
    memo[chosen][prev_word_idx] = result
    return result

def reconstruct(chosen, prev_word_idx, level):
    prev_word = ""
    if level == -1:
        return ""
    if prev_word_idx == -1:
        prev_word = ""
    else:
        prev_word = words[prev_word_idx]
    target = get_min_comb(chosen, prev_word_idx, level)
    for i in range(no_of_word):
        if chosen & 1 << i != 1 << i:
            modified_chosen = chosen | (1 << i);
            combined = combine(prev_word_idx, i)
            if target == len(combined) + get_min_comb(modified_chosen, i, level -1):
                return combined + reconstruct(modified_chosen, i, level-1)

if __name__ == "__main__":
    global no_of_word
    words =[]
    for _ in range(int(sys.stdin.readline().strip())):
        words = []
        memo = [[''] * 15 for _ in range((1 << 16)+ 1)];
        memo2 = [[''] * 15 for _ in range(15)]
        no_of_word = int(sys.stdin.readline().strip());
        for _ in range(no_of_word):
            words.append(sys.stdin.readline().strip())

        to_be_deleted = []
        for i in range(len(words)):
            for j in range(i+1, len(words)):
                if words[i] in words[j]:
                    to_be_deleted.append(i)
                elif words[j] in words[i]:
                    to_be_deleted.append(j)
        for i in to_be_deleted:
            del words[i]
        
        no_of_word = len(words)
        print(reconstruct(1<<15, -1, no_of_word-1))
    